src/Platform/Microsoft.Testing.Platform/OutputDevice/Console/TestProgressFrame.cs

```diff
@@ -0,0 +1,137 @@
+        terminal.Append("✓");
+        terminal.Append(passed.ToString(CultureInfo.CurrentCulture));
+        terminal.ResetColor();
+
+        terminal.Append("/");
+
+        terminal.SetColor(TerminalColor.DarkRed);
+        terminal.Append("x");
+        terminal.Append(failed.ToString(CultureInfo.CurrentCulture));
+        terminal.ResetColor();
+
+        terminal.Append("/");
+
+        terminal.SetColor(TerminalColor.DarkYellow);
+        terminal.Append("?");
+        terminal.Append(skipped.ToString(CultureInfo.CurrentCulture));
+        terminal.ResetColor();
+        terminal.Append(']');
+
+        terminal.Append(' ');
+        terminal.Append(p.AssemblyName);
+        terminal.Append(" (");
+        terminal.Append(p.TargetFramework);
+        terminal.Append('|');
+        terminal.Append(p.Architecture);
+        terminal.Append(')');
+        if (!RoslynString.IsNullOrWhiteSpace(detail))
+        {
+            terminal.Append(" - ");
+            terminal.Append(detail);
+        }
+
+        terminal.SetCursorHorizontal(Width - durationString.Length);
+        terminal.Append(durationString);
+    }
+
+    /// <summary>
+    /// Render VT100 string to update from current to next frame.
+    /// </summary>
+    public void Render(TestProgressFrame previousFrame, AnsiTerminal terminal)
+    {
+        // Move cursor back to 1st line of progress.
+        terminal.MoveCursorUp(previousFrame.ProgressCount + 1);
+
+        int i = 0;
+        for (; i < ProgressCount; i++)
+        {
+            // Do we have previous node string to compare with?
+            if (previousFrame.ProgressCount > i)
+            {
+                if (previousFrame._progressItems[i] == _progressItems[i])
+                {
+                    // Same everything except time, AND same number of digits in time
+                    string durationString = $" ({_progressItems[i].TestWorkerProgress.Stopwatch.Elapsed.TotalSeconds:F1}s)";
+
+                    terminal.SetCursorHorizontal(MaxColumn);
+                    terminal.Append($"{AnsiCodes.SetCursorHorizontal(MaxColumn)}{AnsiCodes.MoveCursorBackward(durationString.Length)}{durationString}");
+                }
+                else
+                {
+                    // TODO: check components to figure out skips and optimize this
+                    terminal.Append($"{AnsiCodes.CSI}{AnsiCodes.EraseInLine}");
+                    AppendTestWorkerProgress(i, terminal);
+                }
+            }
+            else
+            {
+                // From now on we have to simply WriteLine
+                AppendTestWorkerProgress(i, terminal);
+            }
+
+            // Next line
+            terminal.AppendLine();
+        }
+
+        // clear no longer used lines
+        if (i < previousFrame.ProgressCount)
+        {
+            terminal.Append($"{AnsiCodes.CSI}{AnsiCodes.EraseInDisplay}");
+        }
+    }
+
+    public void Clear() => ProgressCount = 0;
```

Comment:
Shall we also reset `_progressItems`?

Reply:
We could, but we will just spend more time cleaning up the array even though we will never touch it again. Keeping as is.